object/option_values.go

```diff
@@ -0,0 +1,203 @@
+// types.BaseOptionValue, such as ExtraConfig.
+type OptionValues []types.BaseOptionValue
+
+// OptionValuesFromMap returns a new OptionValues object from the provided map.
+func OptionValuesFromMap[T any](in map[string]T) OptionValues {
+	if len(in) == 0 {
+		return nil
+	}
+	var (
+		i   int
+		out = make(OptionValues, len(in))
+	)
+	for k, v := range in {
+		out[i] = &types.OptionValue{Key: k, Value: v}
+		i++
+	}
+	return out
+}
+
+// Get returns the value if exists, otherwise nil is returned. The second return
+// value is a flag indicating whether the value exists or nil was the actual
+// value.
+func (ov OptionValues) Get(key string) (any, bool) {
+	if ov == nil {
+		return nil, false
+	}
+	for i := range ov {
+		if optVal := ov[i].GetOptionValue(); optVal != nil {
+			if optVal.Key == key {
+				return optVal.Value, true
+			}
+		}
+	}
+	return nil, false
+}
+
+// GetString returns the value as a string if the value exists.
+func (ov OptionValues) GetString(key string) (string, bool) {
+	if ov == nil {
+		return "", false
+	}
+	for i := range ov {
+		if optVal := ov[i].GetOptionValue(); optVal != nil {
+			if optVal.Key == key {
+				return getOptionValueAsString(optVal.Value), true
+			}
+		}
+	}
+	return "", false
+}
+
+// Additions returns a diff that includes only the elements from the provided
+// list that do not already exist.
+func (ov OptionValues) Additions(in ...types.BaseOptionValue) OptionValues {
+	return ov.diff(in, true)
+}
+
+// Diff returns a diff that includes the elements from the provided list that do
+// not already exist or have different values.
+func (ov OptionValues) Diff(in ...types.BaseOptionValue) OptionValues {
+	return ov.diff(in, false)
+}
+
+func (ov OptionValues) diff(in OptionValues, addOnly bool) OptionValues {
+	if ov == nil && in == nil {
+		return nil
+	}
+	var (
+		out         OptionValues
+		leftOptVals = ov.Map()
+	)
+	for i := range in {
+		if rightOptVal := in[i].GetOptionValue(); rightOptVal != nil {
+			k, v := rightOptVal.Key, rightOptVal.Value
+			if ov == nil {
+				out = append(out, &types.OptionValue{Key: k, Value: v})
+			} else if leftOptVal, ok := leftOptVals[k]; !ok {
+				out = append(out, &types.OptionValue{Key: k, Value: v})
+			} else if !addOnly && v != leftOptVal {
+				out = append(out, &types.OptionValue{Key: k, Value: v})
+			}
+		}
+	}
+	if len(out) == 0 {
+		return nil
+	}
+	return out
+}
+
+// Join combines this list with the provided one and returns the result, joining
+// the two lists on their shared keys.
+// Please note, Join(left, right) means the values from right will be appended
+// to left, without overwriting any values that have shared keys. To overwrite
+// the shared keys in left from right, use Join(right, left) instead.
+func (ov OptionValues) Join(in ...types.BaseOptionValue) OptionValues {
+	var (
+		out     OptionValues
+		outKeys = map[string]struct{}{}
+	)
+
+	// Init the out slice from the left side.
+	if len(ov) > 0 {
```

Comment:
It is, but not as the code was ported. Fixing it now.

Reply:
FYI, this was updated. The check remains, but the map `outKeys` is now only initialized if the left side has any data. Otherwise there is no need to initialize the map.